src/nublado/startup/services/preparer.py

```diff
@@ -0,0 +1,1193 @@
+        if temp_path:
+            self._env[env_v] = str(temp_path)
+            self._logger.debug(f"Set {env_v} to {temp_path!s}")
+            return
+        # In any sane RSP environment, /tmp will not be shared (it will
+        # be either tmpfs or on ephemeral storage, and in any case not
+        # visible beyond its own pod), so we are not actually using a risky
+        # shared directory.
+        self._env[env_v] = "/tmp/butler_cache"
+
+    def _set_cpu_variables(self) -> None:
+        self._logger.debug("Setting CPU threading variables")
+        try:
+            cpu_limit = int(float(self._env.get("CPU_LIMIT", "1")))
+        except ValueError:
+            cpu_limit = 1
+        cpu_limit = max(cpu_limit, 1)
+        cpu_limit_str = str(cpu_limit)
+        for vname in (
+            "CPU_LIMIT",
+            "CPU_COUNT",
+            "GOTO_NUM_THREADS",
+            "MKL_DOMAIN_NUM_THREADS",
+            "MPI_NUM_THREADS",
+            "NUMEXPR_NUM_THREADS",
+            "NUMEXPR_MAX_THREADS",
+            "OMP_NUM_THREADS",
+            "OPENBLAS_NUM_THREADS",
+            "RAYON_NUM_THREADS",
+        ):
+            self._env[vname] = cpu_limit_str
+            self._logger.debug(f"Set '{vname}' -> '{cpu_limit_str}'")
+
+    def _set_image_digest(self) -> None:
+        self._logger.debug("Setting image digest if available")
+        # get_digest() is already a helper function in our parent package.
+        digest = get_digest()
+        if digest:
+            self._logger.debug(f"Set image digest to '{digest}'")
+            self._env["IMAGE_DIGEST"] = digest
+        else:
+            self._logger.debug("Could not get image digest")
+
+    def _expand_panda_tilde(self) -> None:
+        self._logger.debug("Expanding tilde in PANDA_CONFIG_ROOT, if needed")
+        if "PANDA_CONFIG_ROOT" in self._env:
+            # We've already been through set_user(), so USER must be set.
+            username = self._env["USER"]
+            path = Path(self._env["PANDA_CONFIG_ROOT"])
+            path_parts = path.parts
+            if path_parts[0] in ("~", f"~{username}"):
+                new_path = Path(self._home, *path_parts[1:])
+                self._logger.debug(
+                    f"Replacing PANDA_CONFIG_ROOT '{path!s}'"
+                    f"with '{new_path!s}'"
+                )
+                self._env["PANDA_CONFIG_ROOT"] = str(new_path)
+            elif path_parts[0].startswith("~"):
+                self._logger.warning(f"Cannot expand tilde in '{path!s}'")
+
+    async def _set_firefly_variables(self) -> None:
+        self._logger.debug("Setting firefly variables")
+        emsg = "Discovery of portal service failed; not setting variables"
+        try:
+            ff_url = await self._discovery.url_for_ui("portal")
+        except RepertoireError:
+            self._logger.exception(emsg)
+            return
+        finally:
+            await self._discovery.aclose()  # Done with discovery client.
```

Comment:
The Discovery Client is created in the Preparer init method, but then closed here, and I wonder if we'd run into issues in cases where an exception is raised earlier, leading to `_set_firefly_variables` not being called and thus never closing the client?

Maybe here the discover client should just be created here, or the init / aclose should be restructured a bit?

Reply:
Yeah, you will note in a recent lsst.rsp PR there's now a guard around "what if it's already closed"; this all gets confusing with respect to whether this work is coming from an init container or at Lab startup.

Once everything is safely migrated we can clean it up a lot, but I think I like your idea of just creating the client at the point of use